Add a cleanRequests() function to discard previous requests received by ServerMock

loicsimon opened the pull request:
In order to reuse handlers in unit tests without loading a new ServerMock, this function clean the requests stored in ServerMock.

This function is also invoked in `resetHandlers()`.

loicsimon committed:
Add a 'cleanRequest() function.
This function can be invoked explicitly. It is also invoked implicitly by 'resetHandlers()'

loicsimon committed:
Update documentation

lele85 commented:
@loicsimon First of all thanks for the contribution, the PR looks very good and is good point to start.

Before pushing it forward and merging it i would like to understand the use case. Is it performance related?

The next step is then updating the documentation according with the new feature.

loicsimon commented:
> @loicsimon First of all thanks for the contribution, the PR looks very good and is good point to start.
> 
> Before pushing it forward and merging it i would like to understand the use case. Is it performance related?
> 
> The next step is then updating the documentation according with the new feature.

Yes, our problem was related to performance.

We have hundreds of integration tests. Stopping/starting the server before each test was taking too long (cf. #24 ). That's why I added the cleanRequests() method.

I updated README.md. Is there another documentation somewhere?

teorossi82 commented on src/server.js:
```diff
@@ -277,8 +278,13 @@ function Server(host, port, key, cert)
 
     this.resetHandlers = function() {
         handlers = [];
+        self.cleanRequests();
```
I'm not 100% sure in cleaning the history of the requests inside an already existing method called `resetHandlers`, for a couple of reasons:
1 - we are linking cleaning a configuration with getting rid of all the traces left, that could be yet useful
2 - we are changing the behaviour of a method that could potentially breaks past implementations (I'm thinking if someone are using the `resetHandlers` method before doing expectations about requests)

I think we should add another method `reset()` where trigger both `resetHandlers` and `cleanRequests` methods. What do you think @loicsimon and @lele85?

lele85 replied:
Yep @teorossi82, probably having the three methods like this:

- `reset()`
- `resetHandlers()`
- `resetRequests()`

where `reset()` is just `resetHandlers(); resetRequests()` is the best way to go without changing behaviour of existing APIs.

@loicsimon what is your take?

loicsimon replied:
You're right

I don't want to introduce breaking changes for other users.
I'll do my best to update this PR by tomorrow.

lele85 replied:
Thanks!

teorossi82 reviewed (commented):
Thanks @loicsimon! 
I left a comment that I think we should take in consideration before go ahead.

loicsimon committed:
Rename cleanRequests() to resetRequests().
Add reset() method which calls resetHandlers() and resetRequests()

loicsimon commented:
Here is my new commit.

I hope it's alright :)

Thanks for your library! It's really helpful.

teorossi82 commented:
Thanks @loicsimon, code looks good to me now, but please wait @lele85 for approval as well.

teorossi82 reviewed (approved):
Thanks @loicsimon, code looks good to me now, but please wait @lele85 for approval as well.

teorossi82 commented:
@loicsimon thanks for your contribution. I just merged your changes in master and bumped the version. The new package is published in `npm`: https://www.npmjs.com/package/mock-http-server

loicsimon commented:
Thank you. :smile: 

I don't see any new tag and release on github. Is this normal?

teorossi82 commented:
> I don't see any new tag and release on github. Is this normal?

Just created: https://github.com/spreaker/node-mock-http-server/releases/tag/1.4.2
